onnxruntime/core/providers/qnn/qnn_provider_factory.cc

```diff
@@ -93,4 +114,119 @@ ORT_API(onnxruntime::Provider*, GetProvider) {
+    return factory->vendor.c_str();
+  }
+
+  // Creates and returns OrtEpDevice instances for all OrtHardwareDevices that this factory supports.
+  // An EP created with this factory is expected to be able to execute a model with *all* supported
+  // hardware devices at once. A single instance of QNN EP is not currently setup to partition a model among
+  // multiple different QNN backends at once (e.g, npu, cpu, gpu), so this factory instance is set to only
+  // support one backend: npu. To support a different backend, like gpu, create a different factory instance
+  // that only supports GPU.
+  static OrtStatus* GetSupportedDevicesImpl(OrtEpFactory* this_ptr,
+                                            const OrtHardwareDevice* const* devices,
+                                            size_t num_devices,
+                                            OrtEpDevice** ep_devices,
+                                            size_t max_ep_devices,
+                                            size_t* p_num_ep_devices) {
+    size_t& num_ep_devices = *p_num_ep_devices;
+    auto* factory = static_cast<QnnEpFactory*>(this_ptr);
+
+    for (size_t i = 0; i < num_devices && num_ep_devices < max_ep_devices; ++i) {
+      const OrtHardwareDevice& device = *devices[i];
+      if (factory->ort_api.HardwareDevice_Type(&device) == factory->ort_hw_device_type &&
+          factory->ort_api.HardwareDevice_VendorId(&device) == factory->vendor_id) {
+        OrtKeyValuePairs* ep_options = nullptr;
+        factory->ort_api.CreateKeyValuePairs(&ep_options);
+        factory->ort_api.AddKeyValuePair(ep_options, "backend_type", factory->qnn_backend_type.c_str());
+        ORT_API_RETURN_IF_ERROR(
+            factory->ort_api.GetEpApi()->CreateEpDevice(factory, &device, nullptr, ep_options,
+                                                        &ep_devices[num_ep_devices++]));
+      }
+    }
+
+    return nullptr;
+  }
+
+  static OrtStatus* CreateEpImpl(OrtEpFactory* /*this_ptr*/,
+                                 _In_reads_(num_devices) const OrtHardwareDevice* const* /*devices*/,
+                                 _In_reads_(num_devices) const OrtKeyValuePairs* const* /*ep_metadata*/,
+                                 _In_ size_t /*num_devices*/,
+                                 _In_ const OrtSessionOptions* /*session_options*/,
+                                 _In_ const OrtLogger* /*logger*/,
+                                 _Out_ OrtEp** /*ep*/) {
+    return onnxruntime::CreateStatus(ORT_INVALID_ARGUMENT, "QNN EP factory does not support this method.");
+  }
+
+  static void ReleaseEpImpl(OrtEpFactory* /*this_ptr*/, OrtEp* /*ep*/) {
+    // no-op as we never create an EP here.
+  }
+
+  const OrtApi& ort_api;
+  const std::string ep_name;                       // EP name
+  const std::string vendor{"Microsoft"};           // EP vendor name
+  const uint32_t vendor_id{0x4d4f4351};            // Qualcomm vendor ID
```

Comment:
should maybe point out that this isn't an official vendor id. it's an arbitrary value created by bitshifting QCOM. a link to the cpuid_info code would be good. 

https://github.com/microsoft/onnxruntime/blob/6fa8ba107f461d72073ee4842830a9d0a96d2546/onnxruntime/core/common/cpuid_info.cc#L158

but maybe I've screwed that up. I shifted the Q by 24 bits (original code had it as the first byte) thinking that would make the values 0xQCOM but if the vendor_id is 0x4d4f4351 I've not taken into account it's little endian and I've made it 0xMOCQ which is the complete opposite of my intention.

Reply:
But godbolt.org suggests the vendor id with the latest main should be 0x51434f4d which was what I intended.

![image](https://github.com/user-attachments/assets/d10500ef-53f0-4500-986a-4d402eb15819)

Reply:
Interesting. It's not getting the vendor_id from that line in cpuid_info.cc. The vendor is `""`.
<img width="416" alt="image" src="https://github.com/user-attachments/assets/fc2d7b3b-ceb7-4122-84be-a260a577fd0a" />

Let me trace where it is getting the vendor_id from.

Reply:
It's getting the vendor id from dxcore. It is 'M', 'O', 'C', 'Q' = 0x4d4f4351.
<img width="569" alt="image" src="https://github.com/user-attachments/assets/ec827406-538c-4aca-ba36-95469f8ccef7" />

I think I should update the cpuid_info.cc `GetVendorId` to return:
```c++
if (vendor.find("Qualcomm") == 0) return 'M' << 24 | 'O' << 16 | 'C' << 8 | 'Q';
```

Reply:
Yeah, looking around the internet, the correct value is 0x4d4f4351. Also, it makes since in a little-endian machine, since the 'Q' should be the first byte, which means it is **not** shifted.

I'm updating cpuid_info.cc to return `'M' << 24 | 'O' << 16 | 'C' << 8 | 'Q';`

Reply:
Thanks! That needs to match. I didn't realise this was an 'official' id.